Issue/14055 - Mark referrer as spam

nikola-milicevic opened the pull request:
### Mark referrer as spam

Part of [#14055](https://github.com/wordpress-mobile/WordPress-iOS/issues/14055)

This PR includes needed changes to the `StatsTopReferrersTimeIntervalData` and `StatsReferrer`, in order to be able to mark referrers as spam. It does so by adding `isSpam: Bool` to the `StatsReferrer`, and by making `children` and `referrers` properties mutable.
Also, a method is added to the `StatsServiceRemoteV2` for marking / unmarking referrers as spam.

- [x] Please check here if your pull request includes additional test coverage.

nikola-milicevic committed:
Add isSpam property, make children and referrers properties mutable to be able to mark referrer as spam.

nikola-milicevic committed:
Add API call for marking / unmarking referrer as spam.

emilylaguna commented on WordPressKit/StatsServiceRemoteV2.swift:
```diff
@@ -59,6 +59,28 @@ public class StatsServiceRemoteV2: ServiceRemoteWordPressComREST {
+                                failure: @escaping (Error) -> Void) {
+        let path: String
+        if currentValue {
+            path = self.path(forEndpoint: "sites/\(siteID)/stats/referrers/spam/delete?domain=\(referrerDomain)", withVersion: ._1_1)
+        } else {
+            path = self.path(forEndpoint: "sites/\(siteID)/stats/referrers/spam/new?domain=\(referrerDomain)", withVersion: ._1_1)
+        }
```
Since the only difference between the 2 strings is the `delete` and `new`, what do you think about condensing this a bit?

nikola-milicevic replied:
I agree.

emilylaguna commented on WordPressKit/StatsServiceRemoteV2.swift:
```diff
@@ -59,6 +59,28 @@ public class StatsServiceRemoteV2: ServiceRemoteWordPressComREST {
+        }
+
+        wordPressComRestApi.POST(path, parameters: nil) { object, _ in
+            success()
```
Since the API also returns a `success` value in the response what do you think about also validating that before triggering the `success()` callback?

nikola-milicevic replied:
I thought API is designed to return success if the operation was successful, so double checking seemed redundant to me. But if that's not the case (if the success block can be reached even if the operation was unsuccessful) then double checking is a must.

emilylaguna commented on WordPressKit/StatsServiceRemoteV2.swift:
```diff
@@ -59,6 +59,28 @@ public class StatsServiceRemoteV2: ServiceRemoteWordPressComREST {
+    /// - parameters:
+    ///   - referrerDomain: A referrer's domain.
+    ///   - currentValue: Current value of the `isSpam` referrer's property.
+    public func toggleSpamState(for referrerDomain: String,
+                                currentValue: Bool,
+                                success: @escaping () -> Void,
+                                failure: @escaping (Error) -> Void) {
```
I think being more explicit with the methods for the network is better, perhaps 2 methods for mark as spam and unmark as spam would be better. Thoughts?

nikola-milicevic replied:
I agree with the more explicit approach with API calls + deciding logic shouldn't be responsibility of this class, but in this state-changing logic it seems to me it only increases verbosity at the call site, with no little to no upsides.

emilylaguna reviewed (changes requested):
👋 @nikola-milicevic! Left some comments for you!

We should also update the unit tests for the `StatsServiceRemoteV2` class to include the new spam functionality.

nikola-milicevic commented:
> 👋 @nikola-milicevic! Left some comments for you!
> 
> We should also update the unit tests for the `StatsServiceRemoteV2` class to include the new spam functionality.

Thanks for the feedback, if you agree, my plan was to add tests at the very end, when we have everything else in place.

nikola-milicevic committed:
Add MarkAsSpam error type

nikola-milicevic committed:
Add helper method for making API path, add response structure.

nikola-milicevic committed:
Use helper for path creation, perform error checking.

nikola-milicevic committed:
Add response for marking / unmarking referrer as spam.

nikola-milicevic committed:
Add tests for marking / unmarking referrer as spam.

nikola-milicevic committed:
Remove double spacing.

nikola-milicevic commented:
👋 @emilylaguna I updated the code. I added additional error checking as well as tests. Also, I didn't split toggle functionality into two methods, if you think that's better approach - I will refactor it.

nikola-milicevic committed:
Add named parameters

nikola-milicevic committed:
Add named parameters to tests.

emilylaguna commented on WordPressKit/StatsServiceRemoteV2.swift:
```diff
@@ -279,6 +289,26 @@ extension StatsServiceRemoteV2 {
+        return self.path(forEndpoint: "sites/\(siteID)/stats/referrers/spam/\(action)?domain=\(referrerDomain)", withVersion: ._1_1)
+    }
+
+    struct MarkAsSpamResponse {
```
Nice approach, How do you feel about reducing the boiler plate by adhering this to Decodable?

nikola-milicevic replied:
That was my intention, though `POST` method doesn't return `Data`, it already returns a dictionary, so I think that would just introduce unneeded double conversion, if I'm not mistaken :) (from dictionary to `Data`, and then decode to struct).

emilylaguna replied:
Makes sense. For larger objects I think this conversion makes sense but since it's small I don't think it's required.

emilylaguna reviewed (changes requested):
Nice work! I left 1 comment about reducing some boilerplate by using Decodable

emilylaguna reviewed (approved):
LGTM.

nikola-milicevic committed:
Bump podspec version.

nikola-milicevic committed:
Increase access level for StatsServiceRemoteV2 to be able to test it
